// @proappstore/sdk/dist/ai.js
/**
 * Workers AI primitive — server-side LLM + embeddings, included in the
 * platform subscription quota. No per-app key management; the platform
 * pays for the inference and bills usage back via the standard
 * platform-subscription split.
 *
 * @example
 *   const { text } = await app.ai.generate('Write a haiku about yoga');
 *
 *   const result = await app.ai.chat([
 *     { role: 'system', content: 'You are a yoga instructor.' },
 *     { role: 'user', content: 'What is downward dog?' },
 *   ]);
 *
 *   const { vectors } = await app.ai.embed(['vinyasa flow', 'restorative yoga']);
 */
export class AI {
    apiBase;
    auth;
    constructor(apiBase, auth) {
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** Generate text from a single prompt. */
    async generate(prompt, opts = {}) {
        return this.post('/v1/ai/generate', { prompt, ...opts });
    }
    /** Multi-turn chat completion. Pass system + user + assistant messages in order. */
    async chat(messages, opts = {}) {
        return this.post('/v1/ai/generate', { messages, ...opts });
    }
    /** Embed one or many strings. Pass a string for a single embedding, or string[] for a batch. */
    async embed(text, opts = {}) {
        return this.post('/v1/ai/embed', { text, ...opts });
    }
    async post(path, body) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const res = await fetch(this.apiBase + path, {
            method: 'POST',
            headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
            body: JSON.stringify(body),
        });
        if (res.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!res.ok) {
            const text = await res.text().catch(() => '');
            throw new Error(`${path} failed: ${res.status} ${text}`);
        }
        return (await res.json());
    }
}
