// @proappstore/sdk/dist/db.js
import { TenantScope } from './tenant.js';
export class Database {
    appId;
    dataApiBase;
    auth;
    constructor(appId, dataApiBase, auth) {
        this.appId = appId;
        this.dataApiBase = dataApiBase;
        this.auth = auth;
    }
    /** Run a SELECT or other query that returns rows. */
    async query(sql, params) {
        return this.req('/query', { sql, params });
    }
    /** Run an INSERT, UPDATE, DELETE, or DDL statement. */
    async execute(sql, params) {
        return this.req('/execute', { sql, params });
    }
    /** Run multiple statements in a single D1 batch (transactional). */
    async batch(statements) {
        const result = await this.req('/batch', { statements });
        return result.results;
    }
    /**
     * Run named migrations. Each migration runs once — the data-worker tracks
     * which have been applied in a `_migrations` table. Idempotent: safe to
     * call on every app load.
     *
     * @example
     * await app.db.migrate([
     *   { name: '0001_init', sql: 'CREATE TABLE events (id TEXT PRIMARY KEY, ...)' },
     *   { name: '0002_photos', sql: 'ALTER TABLE events ADD COLUMN photo_url TEXT' },
     * ])
     */
    async migrate(migrations) {
        return this.req('/migrate', { migrations });
    }
    /**
     * Return a tenant-scoped wrapper. All `.find`, `.insert`, `.update`, `.delete`
     * calls on the returned scope auto-inject `tenant_id` — the standard way to
     * implement row-level isolation on a shared multi-tenant D1.
     *
     * @example
     *   const tx = app.db.tenant(currentStudio.id);
     *   await tx.insert('clients', { id, name });
     *   const clients = await tx.findMany('clients');
     */
    tenant(tenantId) {
        return new TenantScope(this, tenantId);
    }
    /** List all user-created tables in the database. */
    async tables() {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const response = await fetch(`${this.dataApiBase}/tables`, {
            headers: { Authorization: `Bearer ${token}` },
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok)
            throw new Error(`db.tables failed: ${response.status}`);
        return (await response.json());
    }
    async req(path, body) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const response = await fetch(`${this.dataApiBase}${path}`, {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(body),
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok) {
            const text = await response.text().catch(() => '');
            throw new Error(`db${path} failed: ${response.status} ${text}`);
        }
        return (await response.json());
    }
}
