// @proappstore/sdk/dist/hooks.js
import { useState, useEffect, useCallback, useSyncExternalStore } from 'react';
// ---------------------------------------------------------------------------
// useTheme — vendored from @freeappstore/sdk/hooks
// ---------------------------------------------------------------------------
const THEME_KEY = 'fas:theme';
const themeListeners = new Set();
function getSystemTheme() {
    if (typeof window === 'undefined')
        return 'light';
    return window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
}
function getStoredPreference() {
    if (typeof window === 'undefined')
        return 'system';
    const stored = window.localStorage.getItem(THEME_KEY);
    if (stored === 'light' || stored === 'dark' || stored === 'system')
        return stored;
    return 'system';
}
function resolveTheme(pref) {
    return pref === 'system' ? getSystemTheme() : pref;
}
function applyTheme(theme) {
    if (typeof document === 'undefined')
        return;
    document.documentElement.setAttribute('data-theme', theme);
    document.documentElement.classList.toggle('dark', theme === 'dark');
}
function notifyThemeListeners() {
    for (const fn of themeListeners)
        fn();
}
if (typeof window !== 'undefined') {
    applyTheme(resolveTheme(getStoredPreference()));
    window.matchMedia('(prefers-color-scheme: dark)').addEventListener('change', () => {
        if (getStoredPreference() === 'system') {
            applyTheme(getSystemTheme());
            notifyThemeListeners();
        }
    });
}
function subscribeTheme(cb) {
    themeListeners.add(cb);
    return () => themeListeners.delete(cb);
}
function getThemeSnapshot() {
    const preference = getStoredPreference();
    return { theme: resolveTheme(preference), preference };
}
let cachedSnapshot = getThemeSnapshot();
function getSnapshot() {
    return cachedSnapshot;
}
/**
 * Theme hook — zero-provider. Reads/writes localStorage, applies data-theme on <html>.
 * Shared with FAS SDK (vendored, same localStorage key).
 */
export function useTheme() {
    const snapshot = useSyncExternalStore(subscribeTheme, getSnapshot, getSnapshot);
    const setPreference = useCallback((pref) => {
        if (typeof window === 'undefined')
            return;
        window.localStorage.setItem(THEME_KEY, pref);
        applyTheme(resolveTheme(pref));
        cachedSnapshot = getThemeSnapshot();
        notifyThemeListeners();
    }, []);
    return { theme: snapshot.theme, preference: snapshot.preference, setPreference };
}
/**
 * Auth state + actions. The primary way apps interact with platform identity.
 *
 * Usage:
 * ```tsx
 * const { user, loading, signIn, signOut, deleteAccount } = useProAuth(app)
 * if (loading) return <Spinner />
 * if (!user) return <MySignInPage onSignIn={signIn} />
 * return <MyApp user={user} onSignOut={signOut} />
 * ```
 */
export function useProAuth(app) {
    const [user, setUser] = useState(app.auth.user);
    const [loading, setLoading] = useState(true);
    useEffect(() => {
        app.auth.init().finally(() => setLoading(false));
        return app.auth.onChange(setUser);
    }, [app]);
    const signIn = useCallback(() => app.auth.signIn(), [app]);
    const signOut = useCallback(() => app.auth.signOut(), [app]);
    const deleteAccount = useCallback(async () => {
        try {
            const keys = await app.kv.list();
            for (const key of keys) {
                await app.kv.delete(key).catch(() => { });
            }
        }
        catch { }
        app.auth.signOut();
    }, [app]);
    return { user, loading, signIn, signOut, deleteAccount };
}
/**
 * Subscription state + actions. Check if user is subscribed, upgrade, manage billing.
 *
 * Usage:
 * ```tsx
 * const { subscription, isPro, loading, upgrade, manageBilling } = useProSubscription(app)
 * if (!isPro) return <UpgradePrompt onUpgrade={upgrade} />
 * ```
 */
export function useProSubscription(app) {
    const [subscription, setSubscription] = useState(null);
    const [loading, setLoading] = useState(true);
    useEffect(() => {
        if (!app.auth.token) {
            setLoading(false);
            return;
        }
        app.subscription.status()
            .then(setSubscription)
            .catch(() => { })
            .finally(() => setLoading(false));
    }, [app, app.auth.user]);
    const isPro = subscription?.status === 'active';
    const upgrade = useCallback(async (priceId) => {
        await app.subscription.openCheckout({
            priceId: priceId || 'price_pro_monthly',
            successUrl: window.location.href + (window.location.href.includes('?') ? '&' : '?') + 'upgraded=1',
            cancelUrl: window.location.href,
        });
    }, [app]);
    const manageBilling = useCallback(async () => {
        await app.subscription.openPortal(window.location.href);
    }, [app]);
    return { subscription, isPro, loading, upgrade, manageBilling };
}
/**
 * Push notification state + actions.
 *
 * Usage:
 * ```tsx
 * const { permission, isSubscribed, subscribe, unsubscribe, loading } = useProNotifications(app)
 * if (permission === 'denied') return null
 * return (
 *   <button onClick={isSubscribed ? unsubscribe : subscribe}>
 *     {isSubscribed ? 'Enabled' : 'Enable notifications'}
 *   </button>
 * )
 * ```
 */
export function useProNotifications(app) {
    const [permission, setPermission] = useState(app.notifications.getPermission());
    const [isSubscribed, setIsSubscribed] = useState(false);
    const [loading, setLoading] = useState(true);
    useEffect(() => {
        app.notifications.isSubscribed()
            .then(setIsSubscribed)
            .catch(() => { })
            .finally(() => setLoading(false));
    }, [app]);
    const subscribe = useCallback(async () => {
        setLoading(true);
        try {
            await app.notifications.subscribe();
            setIsSubscribed(true);
            setPermission(app.notifications.getPermission());
        }
        finally {
            setLoading(false);
        }
    }, [app]);
    const unsubscribe = useCallback(async () => {
        setLoading(true);
        try {
            await app.notifications.unsubscribe();
            setIsSubscribed(false);
        }
        finally {
            setLoading(false);
        }
    }, [app]);
    return { permission, isSubscribed, subscribe, unsubscribe, loading };
}
/**
 * Combined auth + subscription gate. Returns the current gate state.
 *
 * Usage:
 * ```tsx
 * const { gate, user, subscription, signIn, upgrade } = useProGate(app)
 * if (gate === 'loading') return <Spinner />
 * if (gate === 'signed-out') return <SignInPage onSignIn={signIn} />
 * if (gate === 'no-subscription') return <UpgradePage onUpgrade={upgrade} />
 * // gate === 'ready' — user is signed in and subscribed
 * return <MyApp user={user!} />
 * ```
 */
export function useProGate(app, opts) {
    const auth = useProAuth(app);
    const sub = useProSubscription(app);
    // allowFree defaults to true while the platform has no payments wired up —
    // see ProShell for the matching default. Flip back when Stripe is live.
    const allowFree = opts?.allowFree ?? true;
    let gate;
    if (auth.loading || (auth.user && sub.loading)) {
        gate = 'loading';
    }
    else if (!auth.user) {
        gate = 'signed-out';
    }
    else if (!allowFree && !sub.isPro) {
        gate = 'no-subscription';
    }
    else {
        gate = 'ready';
    }
    return {
        gate,
        user: auth.user,
        subscription: sub.subscription,
        isPro: sub.isPro,
        signIn: auth.signIn,
        signOut: auth.signOut,
        deleteAccount: auth.deleteAccount,
        upgrade: sub.upgrade,
        manageBilling: sub.manageBilling,
    };
}
