// @proappstore/sdk/dist/index.js
import { FreeAppStore } from '@freeappstore/sdk';
import { AI } from './ai.js';
import { Database } from './db.js';
import { Maps } from './maps.js';
import { Notifications } from './notifications.js';
import { SMS } from './sms.js';
import { Storage } from './storage.js';
import { SubscriptionApi } from './subscription.js';
import { LicenseApi } from './license.js';
import { Usage } from './usage.js';
export { TenantScope } from './tenant.js';
export { Usage } from './usage.js';
/**
 * Pro SDK instance — includes everything from @freeappstore/sdk (auth, kv,
 * counters, rooms, proxy) plus subscription management and license keys.
 *
 * One import, one instance, all features.
 */
export class ProAppStore extends FreeAppStore {
    subscription;
    license;
    db;
    storage;
    maps;
    notifications;
    sms;
    ai;
    usage;
    constructor(opts) {
        super({ appId: opts.appId, ...(opts.fasApiBase && { apiBase: opts.fasApiBase }) });
        const proApiBase = opts.proApiBase ?? 'https://api.proappstore.online';
        this.subscription = new SubscriptionApi(opts.appId, proApiBase, this.auth);
        this.license = new LicenseApi(opts.appId, proApiBase, this.auth);
        this.db = new Database(opts.appId, opts.dataApiBase ?? `https://data-${opts.appId}.proappstore.online`, this.auth);
        this.storage = new Storage(opts.appId, proApiBase, this.auth);
        this.maps = new Maps(proApiBase);
        this.notifications = new Notifications(opts.appId, proApiBase, this.auth);
        this.sms = new SMS(opts.appId, proApiBase, this.auth);
        this.ai = new AI(proApiBase, this.auth);
        this.usage = new Usage(opts.appId, proApiBase, this.auth);
        // Auto-start telemetry unless the app opts out. `start()` is a no-op in
        // SSR / non-browser contexts, so this is safe even when the SDK is
        // imported on the server side.
        if (opts.usage?.auto !== false)
            this.usage.start();
    }
}
/** Create a new ProAppStore SDK instance. Includes all free + pro features. */
export function initPro(opts) {
    return new ProAppStore(opts);
}
