// @proappstore/sdk/dist/license.js
export class LicenseApi {
    appId;
    apiBase;
    auth;
    constructor(appId, apiBase, auth) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** Returns the license info for the signed-in user, or null. */
    async current() {
        const token = this.auth.token;
        if (!token)
            return null;
        const response = await fetch(new URL(`/v1/apps/${encodeURIComponent(this.appId)}/license`, this.apiBase), { headers: { Authorization: `Bearer ${token}` } });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            return null;
        }
        if (response.status === 404)
            return null;
        if (!response.ok)
            throw new Error(`license.current failed: ${response.status}`);
        return (await response.json());
    }
    /** Validate an arbitrary license key against the server (no auth required). */
    async validate(key) {
        const response = await fetch(new URL('/v1/license/validate', this.apiBase), {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ appId: this.appId, key }),
        });
        if (!response.ok)
            return false;
        const { valid } = (await response.json());
        return valid;
    }
}
