// @proappstore/sdk/dist/maps.js
/**
 * Maps — geocoding, reverse geocoding, and map embed helpers.
 * Backed by Nominatim (OpenStreetMap) proxied through the PAS API.
 * No Google API keys needed.
 */
export class Maps {
    apiBase;
    constructor(apiBase) {
        this.apiBase = apiBase;
    }
    /** Convert an address string to coordinates. Returns up to `limit` results. */
    async geocode(query, limit = 5, opts) {
        const url = new URL(`${this.apiBase}/v1/maps/geocode`);
        url.searchParams.set('q', query);
        url.searchParams.set('limit', String(limit));
        const response = await fetch(url.toString(), opts?.signal ? { signal: opts.signal } : undefined);
        if (!response.ok)
            throw new Error(`maps.geocode failed: ${response.status}`);
        const data = (await response.json());
        return data.results;
    }
    /**
     * Driving route between two points. Returns GeoJSON LineString geometry
     * plus distance and duration. Backed by OSRM (OpenStreetMap), proxied
     * through the platform — no API key needed.
     */
    async route(from, to, opts) {
        const url = new URL(`${this.apiBase}/v1/maps/route`);
        url.searchParams.set('from', `${from.lat},${from.lng}`);
        url.searchParams.set('to', `${to.lat},${to.lng}`);
        const response = await fetch(url.toString(), opts?.signal ? { signal: opts.signal } : undefined);
        if (!response.ok)
            throw new Error(`maps.route failed: ${response.status}`);
        return (await response.json());
    }
    /** Convert coordinates to an address. */
    async reverseGeocode(lat, lng, opts) {
        const url = new URL(`${this.apiBase}/v1/maps/reverse`);
        url.searchParams.set('lat', String(lat));
        url.searchParams.set('lng', String(lng));
        const response = await fetch(url.toString(), opts?.signal ? { signal: opts.signal } : undefined);
        if (!response.ok)
            throw new Error(`maps.reverseGeocode failed: ${response.status}`);
        return (await response.json());
    }
    /**
     * Get an OpenStreetMap embed URL for use in an iframe.
     * No API key needed. Free forever.
     *
     * Usage: <iframe src={app.maps.embedUrl(lat, lng, zoom)} />
     */
    embedUrl(lat, lng, zoom = 15) {
        return `https://www.openstreetmap.org/export/embed.html?bbox=${lng - 0.01},${lat - 0.01},${lng + 0.01},${lat + 0.01}&layer=mapnik&marker=${lat},${lng}`;
    }
    /**
     * Get a static map image URL (no JavaScript needed).
     * Uses OpenStreetMap tile server.
     *
     * Usage: <img src={app.maps.staticUrl(lat, lng)} />
     */
    staticUrl(lat, lng, zoom = 15) {
        const x = Math.floor(((lng + 180) / 360) * Math.pow(2, zoom));
        const y = Math.floor((1 - Math.log(Math.tan((lat * Math.PI) / 180) + 1 / Math.cos((lat * Math.PI) / 180)) / Math.PI) / 2 * Math.pow(2, zoom));
        return `https://tile.openstreetmap.org/${zoom}/${x}/${y}.png`;
    }
}
