// @proappstore/sdk/dist/shell.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, useEffect, useCallback } from 'react';
import { ProfileMenu, SignInButton, ProBadge, GateScreen } from './ui.js';
/**
 * ProShell — platform-level Shell for all ProAppStore apps.
 *
 * Handles:
 * - Auth initialization + sign-in gate
 * - Subscription check + upgrade wall (unless allowFree=true)
 * - Topbar with avatar, app name, menu (sign out, delete account, manage billing)
 * - Theme support via CSS custom properties
 * - Only renders children when all gates pass
 *
 * Usage:
 * ```tsx
 * import { initPro } from '@proappstore/sdk'
 * import { ProShell } from '@proappstore/sdk/shell'
 *
 * const app = initPro({ appId: 'meetup' })
 *
 * export default function App() {
 *   return (
 *     <ProShell app={app} appName="Meetup">
 *       <MeetupApp />
 *     </ProShell>
 *   )
 * }
 * ```
 */
export function ProShell({ app, children, appName, allowFree = true, showThemeToggle = true }) {
    const [user, setUser] = useState(app.auth.user);
    const [subscription, setSubscription] = useState(null);
    const [gate, setGate] = useState('loading');
    useEffect(() => {
        app.auth.init();
        return app.auth.onChange((u) => {
            setUser(u);
            if (!u)
                setGate('signed-out');
        });
    }, [app]);
    // Check subscription after auth.
    useEffect(() => {
        if (!user)
            return;
        app.subscription.status().then((sub) => {
            setSubscription(sub);
            if (allowFree || sub?.status === 'active') {
                setGate('ready');
            }
            else {
                setGate('no-subscription');
            }
        }).catch(() => {
            setSubscription(null);
            setGate(allowFree ? 'ready' : 'no-subscription');
        });
    }, [user, app, allowFree]);
    // --- Gates ---
    if (gate !== 'ready') {
        return _jsx(GateScreen, { gate: gate, app: app, appName: appName });
    }
    // --- Ready: render app with topbar ---
    return (_jsxs("div", { style: styles.shell, children: [_jsxs("header", { style: styles.topbar, children: [_jsxs("div", { style: styles.topbarLeft, children: [_jsx("a", { href: "https://proappstore.online", style: styles.logoLink, children: "Pro" }), appName && _jsx("span", { style: styles.appName, children: appName }), subscription?.status === 'active' && _jsx(ProBadge, {})] }), _jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.5rem' }, children: [!user && _jsx(SignInButton, { app: app, label: "Sign in" }), user && _jsx(ProfileMenu, { app: app, showThemeToggle: showThemeToggle })] })] }), _jsx("main", { style: styles.main, children: children }), _jsxs("footer", { style: styles.footer, children: ["Part of", ' ', _jsx("a", { href: "https://proappstore.online", style: { color: 'var(--accent, #7c3aed)', fontWeight: 600, textDecoration: 'none' }, children: "ProAppStore" })] })] }));
}
const styles = {
    shell: { minHeight: '100dvh', display: 'flex', flexDirection: 'column' },
    topbar: {
        display: 'flex', alignItems: 'center', justifyContent: 'space-between',
        padding: '0.5rem 1rem',
        borderBottom: '1px solid var(--border, #e2e8f0)',
        background: 'var(--surface, #ffffff)',
        position: 'sticky', top: 0, zIndex: 50,
    },
    topbarLeft: { display: 'flex', alignItems: 'center', gap: '0.75rem' },
    logoLink: { fontWeight: 800, fontSize: '1rem', color: 'var(--accent, #7c3aed)', textDecoration: 'none' },
    appName: { fontSize: '0.85rem', fontWeight: 600, color: 'var(--muted, #64748b)' },
    main: { flex: 1, display: 'flex', flexDirection: 'column' },
    footer: {
        padding: '1rem', textAlign: 'center', fontSize: '0.75rem',
        color: 'var(--muted, #64748b)',
        borderTop: '1px solid var(--border, #e2e8f0)',
    },
};
