// @proappstore/sdk/dist/storage.js
/**
 * File storage — upload, download, list, delete files.
 * Backed by R2 on the PAS API, scoped to (appId, userId).
 */
export class Storage {
    appId;
    apiBase;
    auth;
    constructor(appId, apiBase, auth) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** Upload a file. Returns the upload result with the file URL. */
    async upload(path, data, contentType) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = `${this.apiBase}/v1/apps/${encodeURIComponent(this.appId)}/storage/${path}`;
        const response = await fetch(url, {
            method: 'PUT',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': contentType || (data instanceof Blob ? data.type : 'application/octet-stream'),
            },
            body: data,
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok) {
            const text = await response.text();
            throw new Error(`storage.upload failed (${response.status}): ${text}`);
        }
        return (await response.json());
    }
    /** Download a file. Returns the Response (use .blob(), .arrayBuffer(), etc.). */
    async download(path) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = `${this.apiBase}/v1/apps/${encodeURIComponent(this.appId)}/storage/${path}`;
        const response = await fetch(url, {
            headers: { Authorization: `Bearer ${token}` },
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (response.status === 404)
            throw new Error('File not found.');
        if (!response.ok)
            throw new Error(`storage.download failed: ${response.status}`);
        return response;
    }
    /**
     * Upload a public file. Anyone can view it without auth (for profile pics, event photos, etc.).
     * Stored under the _public/ prefix. Returns a public URL usable in <img src>.
     */
    async uploadPublic(path, data, contentType) {
        return this.upload(`_public/${path}`, data, contentType);
    }
    /** Get a public URL for a file (no auth needed, usable in <img src>). File must have been uploaded with uploadPublic(). */
    publicUrl(path) {
        return `${this.apiBase}/v1/apps/${encodeURIComponent(this.appId)}/public/${path}`;
    }
    /** Get a private URL for a file (requires auth header). */
    url(path) {
        return `${this.apiBase}/v1/apps/${encodeURIComponent(this.appId)}/storage/${path}`;
    }
    /** List all files for the current user in this app. */
    async list() {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = `${this.apiBase}/v1/apps/${encodeURIComponent(this.appId)}/files`;
        const response = await fetch(url, {
            headers: { Authorization: `Bearer ${token}` },
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok)
            throw new Error(`storage.list failed: ${response.status}`);
        const data = (await response.json());
        return data.files;
    }
    /** Delete a file. */
    async delete(path) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = `${this.apiBase}/v1/apps/${encodeURIComponent(this.appId)}/storage/${path}`;
        const response = await fetch(url, {
            method: 'DELETE',
            headers: { Authorization: `Bearer ${token}` },
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok && response.status !== 404) {
            throw new Error(`storage.delete failed: ${response.status}`);
        }
    }
}
