// @proappstore/sdk/dist/subscription.js
export class SubscriptionApi {
    appId;
    apiBase;
    auth;
    constructor(appId, apiBase, auth) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** Returns the user's current subscription, or null if they have none. */
    async status() {
        const response = await this.req('GET', '/v1/subscription');
        if (response.status === 404)
            return null;
        if (!response.ok)
            throw new Error(`subscription.status failed: ${response.status}`);
        return (await response.json());
    }
    /** Redirects to Stripe-hosted checkout. Page navigates away. */
    async openCheckout(req) {
        const response = await this.req('POST', '/v1/checkout', req);
        if (!response.ok)
            throw new Error(`subscription.openCheckout failed: ${response.status}`);
        const { url } = (await response.json());
        if (typeof window !== 'undefined')
            window.location.assign(url);
    }
    /** Redirects to Stripe customer portal to manage billing. */
    async openPortal(returnUrl) {
        const response = await this.req('POST', '/v1/portal', { returnUrl });
        if (!response.ok)
            throw new Error(`subscription.openPortal failed: ${response.status}`);
        const { url } = (await response.json());
        if (typeof window !== 'undefined')
            window.location.assign(url);
    }
    async req(method, path, body) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const init = {
            method,
            headers: {
                Authorization: `Bearer ${token}`,
                ...(body !== undefined ? { 'Content-Type': 'application/json' } : {}),
            },
        };
        if (body !== undefined)
            init.body = JSON.stringify(body);
        const response = await fetch(new URL(path, this.apiBase), init);
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        return response;
    }
}
