// @proappstore/sdk/dist/ui.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/**
 * @proappstore/sdk/ui — Full UI component library for Pro apps.
 *
 * Re-exports base primitives (Avatar, SignInButton, ThemeToggle, ProfileMenu)
 * plus Pro-specific components: ProBadge, UpgradeCard, BillingButton,
 * SubscriptionStatus, ProProfilePage.
 */
import { useState, useEffect, useCallback, useRef } from 'react';
import { useProAuth, useProSubscription, useTheme } from './hooks.js';
// Re-export base primitives
export { Avatar, ThemeToggle } from './ui-primitives.js';
export function SignInButton({ app, label = 'Sign in with GitHub' }) {
    return (_jsx("button", { onClick: () => app.auth.signIn(), style: {
            background: 'var(--accent, #7c3aed)',
            color: '#fff',
            border: 'none',
            padding: '0.6rem 1.5rem',
            borderRadius: 'var(--radius, 0.75rem)',
            fontSize: '0.9rem',
            fontWeight: 700,
            cursor: 'pointer',
            fontFamily: 'inherit',
        }, children: label }));
}
/** Purple "PRO" badge. Shows subscription tier. */
export function ProBadge({ size = 'sm' }) {
    const sizes = {
        sm: { fontSize: '0.6rem', padding: '0.1rem 0.4rem' },
        md: { fontSize: '0.7rem', padding: '0.2rem 0.5rem' },
        lg: { fontSize: '0.8rem', padding: '0.25rem 0.6rem' },
    };
    return (_jsx("span", { style: {
            ...sizes[size],
            fontWeight: 700,
            background: 'var(--accent, #7c3aed)',
            color: '#fff',
            borderRadius: '0.25rem',
            textTransform: 'uppercase',
            letterSpacing: '0.03em',
            whiteSpace: 'nowrap',
            display: 'inline-flex',
            alignItems: 'center',
        }, children: "PRO" }));
}
/** Inline subscription status: PRO badge or "Free" with optional upgrade link. */
export function SubscriptionStatus({ app, showUpgrade = true }) {
    const { subscription, isPro, loading, upgrade } = useProSubscription(app);
    if (loading) {
        return _jsx("span", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)' }, children: "..." });
    }
    if (isPro) {
        const renewal = subscription?.cancelAtPeriodEnd
            ? 'Cancels at period end'
            : subscription?.currentPeriodEnd
                ? `Renews ${new Date(subscription.currentPeriodEnd).toLocaleDateString()}`
                : '';
        return (_jsxs("span", { style: { display: 'inline-flex', alignItems: 'center', gap: '0.4rem' }, children: [_jsx(ProBadge, { size: "md" }), renewal && _jsx("span", { style: { fontSize: '0.75rem', color: 'var(--muted, #64748b)' }, children: renewal })] }));
    }
    return (_jsxs("span", { style: { display: 'inline-flex', alignItems: 'center', gap: '0.5rem' }, children: [_jsx("span", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)' }, children: "Free plan" }), showUpgrade && (_jsx("button", { onClick: () => upgrade(), style: {
                    background: 'none',
                    border: '1px solid var(--accent, #7c3aed)',
                    color: 'var(--accent, #7c3aed)',
                    padding: '0.2rem 0.6rem',
                    borderRadius: 'var(--radius-sm, 0.5rem)',
                    fontSize: '0.75rem',
                    fontWeight: 600,
                    cursor: 'pointer',
                    fontFamily: 'inherit',
                }, children: "Upgrade" }))] }));
}
/** Styled card prompting the user to upgrade to Pro. */
export function UpgradeCard({ app, title = 'Upgrade to Pro', description = 'Unlock all premium features with a Pro subscription.', priceLabel = '$9/month', features = ['Cloud sync across devices', 'AI-powered features', 'Unlimited storage', 'Priority support'], }) {
    const { upgrade } = useProSubscription(app);
    return (_jsxs("div", { style: {
            background: 'var(--surface, #ffffff)',
            border: '1px solid var(--accent, #7c3aed)',
            borderRadius: 'var(--radius, 0.75rem)',
            padding: '1.75rem',
            maxWidth: 400,
        }, children: [_jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.5rem', marginBottom: '0.5rem' }, children: [_jsx("h3", { style: { fontSize: '1.25rem', fontWeight: 800, color: 'var(--ink-strong, var(--ink, #0f172a))', margin: 0 }, children: title }), _jsx(ProBadge, { size: "md" })] }), _jsx("p", { style: { fontSize: '0.9rem', color: 'var(--muted, #64748b)', marginBottom: '1rem' }, children: description }), features.length > 0 && (_jsx("ul", { style: { listStyle: 'none', padding: 0, margin: '0 0 1.25rem' }, children: features.map((f, i) => (_jsxs("li", { style: {
                        padding: '0.35rem 0',
                        fontSize: '0.85rem',
                        color: 'var(--ink, #1e293b)',
                        borderBottom: '1px solid var(--border, #e2e8f0)',
                        display: 'flex',
                        alignItems: 'center',
                        gap: '0.5rem',
                    }, children: [_jsx("span", { style: { color: 'var(--accent, #7c3aed)', fontWeight: 700 }, children: "+" }), f] }, i))) })), _jsxs("button", { onClick: () => upgrade(), style: {
                    width: '100%',
                    background: 'var(--accent, #7c3aed)',
                    color: '#fff',
                    border: 'none',
                    padding: '0.75rem',
                    borderRadius: 'var(--radius, 0.75rem)',
                    fontSize: '0.9rem',
                    fontWeight: 700,
                    cursor: 'pointer',
                    fontFamily: 'inherit',
                }, children: ["Subscribe \u2014 ", priceLabel] })] }));
}
/** Button that opens the Stripe billing portal. */
export function BillingButton({ app, label = 'Manage billing', variant = 'secondary' }) {
    const [loading, setLoading] = useState(false);
    const handleClick = useCallback(async () => {
        setLoading(true);
        try {
            await app.subscription.openPortal(window.location.href);
        }
        catch { }
        finally {
            setLoading(false);
        }
    }, [app]);
    const baseStyle = {
        padding: '0.6rem 1.25rem',
        borderRadius: 'var(--radius, 0.75rem)',
        fontSize: '0.85rem',
        fontWeight: 600,
        cursor: loading ? 'wait' : 'pointer',
        fontFamily: 'inherit',
        opacity: loading ? 0.7 : 1,
    };
    const variants = {
        primary: { ...baseStyle, background: 'var(--accent, #7c3aed)', color: '#fff', border: 'none' },
        secondary: { ...baseStyle, background: 'transparent', color: 'var(--ink, #1e293b)', border: '1px solid var(--border, #e2e8f0)' },
        ghost: { ...baseStyle, background: 'none', color: 'var(--accent, #7c3aed)', border: 'none', padding: '0.4rem 0.75rem' },
    };
    return (_jsx("button", { onClick: handleClick, disabled: loading, style: variants[variant], children: loading ? 'Opening...' : label }));
}
/** Avatar button that opens dropdown with Pro features: badge, billing, theme, sign out. */
export function ProfileMenu({ app, showThemeToggle = true, showBilling = true, children }) {
    const { user, signOut, deleteAccount } = useProAuth(app);
    const { subscription, isPro, manageBilling } = useProSubscription(app);
    const [open, setOpen] = useState(false);
    const ref = useRef(null);
    useEffect(() => {
        if (!open)
            return;
        const handler = (e) => {
            if (ref.current && !ref.current.contains(e.target))
                setOpen(false);
        };
        document.addEventListener('mousedown', handler);
        return () => document.removeEventListener('mousedown', handler);
    }, [open]);
    if (!user)
        return null;
    const handleSignOut = () => { signOut(); setOpen(false); };
    const handleBilling = async () => { await manageBilling(); setOpen(false); };
    const handleDelete = async () => {
        const subClause = subscription?.status === 'active' ? ' and cancels your subscription' : '';
        if (!confirm(`Delete your account? This permanently removes ALL your data across ALL apps${subClause}. This cannot be undone.`))
            return;
        if (!confirm('Are you absolutely sure? Last chance.'))
            return;
        await deleteAccount();
        setOpen(false);
    };
    return (_jsxs("div", { ref: ref, style: { position: 'relative' }, children: [_jsx("button", { onClick: () => setOpen(!open), style: {
                    background: 'none',
                    border: '2px solid var(--border, #e2e8f0)',
                    borderRadius: '50%',
                    padding: 0,
                    cursor: 'pointer',
                    width: 32,
                    height: 32,
                    overflow: 'hidden',
                    display: 'block',
                }, children: user.avatarUrl ? (_jsx("img", { src: user.avatarUrl, alt: user.login, width: 28, height: 28, style: { borderRadius: '50%', display: 'block' } })) : (_jsx("div", { style: { width: 28, height: 28, borderRadius: '50%', background: 'var(--accent, #7c3aed)', color: '#fff', display: 'flex', alignItems: 'center', justifyContent: 'center', fontSize: 13, fontWeight: 700 }, children: user.login.charAt(0).toUpperCase() })) }), open && (_jsxs("div", { style: {
                    position: 'absolute', top: 40, right: 0,
                    background: 'var(--surface, #ffffff)',
                    border: '1px solid var(--border, #e2e8f0)',
                    borderRadius: 'var(--radius, 0.75rem)',
                    boxShadow: '0 4px 20px rgba(0,0,0,0.1)',
                    minWidth: 220, padding: '0.5rem 0', zIndex: 100,
                }, children: [_jsxs("div", { style: { padding: '0.5rem 1rem', borderBottom: '1px solid var(--border, #e2e8f0)', display: 'flex', alignItems: 'center', gap: '0.5rem' }, children: [_jsx("strong", { style: { fontSize: '0.85rem', color: 'var(--ink, #1e293b)' }, children: user.login }), isPro && _jsx(ProBadge, {})] }), showThemeToggle && (_jsxs("div", { style: { padding: '0.5rem 1rem', borderBottom: '1px solid var(--border, #e2e8f0)', display: 'flex', alignItems: 'center', justifyContent: 'space-between' }, children: [_jsx("span", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)' }, children: "Theme" }), _jsx(ThemeToggleInline, {})] })), showBilling && isPro && (_jsx("button", { onClick: handleBilling, style: menuItemStyle, children: "Manage billing" })), children, _jsx("button", { onClick: handleSignOut, style: menuItemStyle, children: "Sign out" }), _jsx("button", { onClick: handleDelete, style: { ...menuItemStyle, color: '#dc2626' }, children: "Delete account" })] }))] }));
}
/** Inline theme toggle for menu (imports from same file to avoid circular) */
function ThemeToggleInline() {
    const { theme, preference, setPreference } = useTheme();
    const cycle = useCallback(() => {
        const order = ['system', 'light', 'dark'];
        const idx = order.indexOf(preference);
        setPreference(order[(idx + 1) % order.length]);
    }, [preference, setPreference]);
    const icon = theme === 'dark' ? (_jsxs("svg", { width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round", children: [_jsx("circle", { cx: "12", cy: "12", r: "5" }), _jsx("line", { x1: "12", y1: "1", x2: "12", y2: "3" }), _jsx("line", { x1: "12", y1: "21", x2: "12", y2: "23" }), _jsx("line", { x1: "4.22", y1: "4.22", x2: "5.64", y2: "5.64" }), _jsx("line", { x1: "18.36", y1: "18.36", x2: "19.78", y2: "19.78" }), _jsx("line", { x1: "1", y1: "12", x2: "3", y2: "12" }), _jsx("line", { x1: "21", y1: "12", x2: "23", y2: "12" }), _jsx("line", { x1: "4.22", y1: "19.78", x2: "5.64", y2: "18.36" }), _jsx("line", { x1: "18.36", y1: "5.64", x2: "19.78", y2: "4.22" })] })) : (_jsx("svg", { width: "16", height: "16", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round", children: _jsx("path", { d: "M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z" }) }));
    return (_jsx("button", { onClick: cycle, "aria-label": `Theme: ${preference}`, style: {
            width: 32, height: 32, borderRadius: 'var(--radius-sm, 0.5rem)',
            border: '1px solid var(--border, #e2e8f0)', background: 'var(--surface, #ffffff)',
            color: 'var(--ink, #1e293b)', display: 'inline-flex', alignItems: 'center',
            justifyContent: 'center', cursor: 'pointer', padding: 0, fontFamily: 'inherit',
        }, children: icon }));
}
const menuItemStyle = {
    display: 'block', width: '100%', padding: '0.5rem 1rem',
    background: 'none', border: 'none', textAlign: 'left',
    fontSize: '0.85rem', cursor: 'pointer',
    color: 'var(--ink, #1e293b)', fontFamily: 'inherit',
};
/** Full-page profile/settings with subscription info, billing, theme, danger zone. */
export function ProProfilePage({ app, showThemeToggle = true }) {
    const { user, loading, signOut, deleteAccount } = useProAuth(app);
    const { subscription, isPro, loading: subLoading, upgrade, manageBilling } = useProSubscription(app);
    const { preference, setPreference } = useTheme();
    if (loading) {
        return _jsx("div", { style: { padding: '2rem', textAlign: 'center', color: 'var(--muted, #64748b)' }, children: "Loading..." });
    }
    if (!user) {
        return (_jsxs("div", { style: { padding: '2rem', textAlign: 'center' }, children: [_jsx("p", { style: { color: 'var(--muted, #64748b)', marginBottom: '1rem' }, children: "Sign in to view your profile." }), _jsx(SignInButton, { app: app })] }));
    }
    const handleDelete = async () => {
        const subClause = subscription?.status === 'active' ? ' and cancels your subscription' : '';
        if (!confirm(`Delete your account? This permanently removes ALL your data across ALL apps${subClause}. This cannot be undone.`))
            return;
        if (!confirm('Are you absolutely sure? Last chance.'))
            return;
        await deleteAccount();
    };
    const themeOptions = [
        { value: 'system', label: 'System' },
        { value: 'light', label: 'Light' },
        { value: 'dark', label: 'Dark' },
    ];
    return (_jsxs("div", { style: { maxWidth: 520, margin: '0 auto', padding: '2rem 1rem' }, children: [_jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '1rem', marginBottom: '2rem' }, children: [user.avatarUrl ? (_jsx("img", { src: user.avatarUrl, alt: user.login, width: 64, height: 64, style: { borderRadius: '50%' } })) : (_jsx("div", { style: { width: 64, height: 64, borderRadius: '50%', background: 'var(--accent, #7c3aed)', color: '#fff', display: 'flex', alignItems: 'center', justifyContent: 'center', fontSize: 28, fontWeight: 700 }, children: user.login.charAt(0).toUpperCase() })), _jsxs("div", { children: [_jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.5rem' }, children: [_jsx("span", { style: { fontSize: '1.25rem', fontWeight: 700, color: 'var(--ink-strong, var(--ink, #0f172a))' }, children: user.login }), isPro && _jsx(ProBadge, { size: "md" })] }), _jsx("div", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)' }, children: "ProAppStore account" })] })] }), _jsx(Section, { title: "Subscription", children: subLoading ? (_jsx("p", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)' }, children: "Loading..." })) : isPro ? (_jsxs("div", { children: [_jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.5rem', marginBottom: '0.75rem' }, children: [_jsx(ProBadge, { size: "lg" }), _jsx("span", { style: { fontSize: '0.9rem', fontWeight: 600, color: 'var(--ink, #1e293b)' }, children: "Active" })] }), subscription?.currentPeriodEnd && (_jsxs("p", { style: { fontSize: '0.82rem', color: 'var(--muted, #64748b)', marginBottom: '0.75rem' }, children: [subscription.cancelAtPeriodEnd ? 'Cancels' : 'Renews', " on ", new Date(subscription.currentPeriodEnd).toLocaleDateString()] })), _jsx(BillingButton, { app: app, variant: "secondary" })] })) : (_jsxs("div", { children: [_jsx("p", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)', marginBottom: '0.75rem' }, children: "You're on the free plan. Upgrade to unlock all premium features." }), _jsx("button", { onClick: () => upgrade(), style: {
                                background: 'var(--accent, #7c3aed)', color: '#fff', border: 'none',
                                padding: '0.6rem 1.25rem', borderRadius: 'var(--radius, 0.75rem)',
                                fontSize: '0.85rem', fontWeight: 700, cursor: 'pointer', fontFamily: 'inherit',
                            }, children: "Upgrade to Pro \u2014 $9/mo" })] })) }), showThemeToggle && (_jsx(Section, { title: "Appearance", children: _jsx("div", { style: { display: 'flex', gap: '0.5rem' }, children: themeOptions.map((opt) => (_jsx("button", { onClick: () => setPreference(opt.value), style: {
                            flex: 1, padding: '0.5rem',
                            borderRadius: 'var(--radius-sm, 0.5rem)',
                            border: preference === opt.value ? '2px solid var(--accent, #7c3aed)' : '1px solid var(--border, #e2e8f0)',
                            background: preference === opt.value ? 'var(--accent-soft, #f5f3ff)' : 'transparent',
                            color: preference === opt.value ? 'var(--accent, #7c3aed)' : 'var(--muted, #64748b)',
                            fontWeight: preference === opt.value ? 700 : 500,
                            fontSize: '0.85rem', cursor: 'pointer', fontFamily: 'inherit',
                        }, children: opt.label }, opt.value))) }) })), _jsx("button", { onClick: signOut, style: {
                    width: '100%', padding: '0.75rem',
                    borderRadius: 'var(--radius, 0.75rem)',
                    border: '1px solid var(--border, #e2e8f0)',
                    background: 'var(--surface, #ffffff)',
                    color: 'var(--ink, #1e293b)',
                    fontSize: '0.9rem', fontWeight: 600,
                    cursor: 'pointer', marginBottom: '1.5rem', fontFamily: 'inherit',
                }, children: "Sign out" }), _jsxs("div", { style: { border: '1px solid #fecaca', borderRadius: 'var(--radius, 0.75rem)', padding: '1.25rem' }, children: [_jsx("div", { style: { fontSize: '0.9rem', fontWeight: 700, color: '#dc2626', marginBottom: '0.5rem' }, children: "Danger zone" }), _jsxs("p", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)', marginBottom: '0.75rem' }, children: ["Permanently delete your account and all data across all apps.", isPro && ' Your subscription will be cancelled.'] }), _jsx("button", { onClick: handleDelete, style: {
                            padding: '0.5rem 1rem', borderRadius: 'var(--radius-sm, 0.5rem)',
                            border: '1px solid #dc2626', background: 'transparent',
                            color: '#dc2626', fontSize: '0.85rem', fontWeight: 600,
                            cursor: 'pointer', fontFamily: 'inherit',
                        }, children: "Delete account" })] })] }));
}
/** Renders the appropriate gate screen (loading, sign-in, or upgrade). */
export function GateScreen({ gate, app, appName }) {
    if (gate === 'loading') {
        return (_jsx("div", { style: { minHeight: '100dvh', display: 'flex', alignItems: 'center', justifyContent: 'center' }, children: _jsx("p", { style: { color: 'var(--muted, #64748b)' }, children: "Loading..." }) }));
    }
    if (gate === 'signed-out') {
        return (_jsx("div", { style: { minHeight: '100dvh', display: 'flex', alignItems: 'center', justifyContent: 'center', padding: '2rem' }, children: _jsxs("div", { style: { maxWidth: 400, textAlign: 'center' }, children: [_jsx("h1", { style: { fontSize: '1.5rem', fontWeight: 800, marginBottom: '0.5rem', color: 'var(--ink, #1e293b)' }, children: appName || 'ProAppStore' }), _jsx("p", { style: { color: 'var(--muted, #64748b)', fontSize: '0.9rem', marginBottom: '1rem' }, children: "Sign in to your ProAppStore account to continue." }), _jsx(SignInButton, { app: app }), _jsx("p", { style: { color: 'var(--muted, #64748b)', fontSize: '0.75rem', marginTop: '0.75rem' }, children: "One account for all Pro apps." })] }) }));
    }
    // no-subscription
    return (_jsx("div", { style: { minHeight: '100dvh', display: 'flex', alignItems: 'center', justifyContent: 'center', padding: '2rem' }, children: _jsx(UpgradeCard, { app: app, title: "Pro subscription required", description: `${appName || 'This app'} requires an active ProAppStore subscription.` }) }));
}
// ---------------------------------------------------------------------------
// Section helper (internal)
// ---------------------------------------------------------------------------
function Section({ title, children }) {
    return (_jsxs("div", { style: {
            background: 'var(--surface, #ffffff)',
            border: '1px solid var(--border, #e2e8f0)',
            borderRadius: 'var(--radius, 0.75rem)',
            padding: '1.25rem',
            marginBottom: '1rem',
        }, children: [_jsx("div", { style: { fontSize: '0.9rem', fontWeight: 700, marginBottom: '0.75rem', color: 'var(--ink, #1e293b)' }, children: title }), children] }));
}
