// @proappstore/sdk/dist/usage.js
/**
 * Usage telemetry — heartbeats `POST /v1/usage/ping` while the tab is visible.
 *
 * Auto-started by `initPro()` unless `usage.auto === false` in the options.
 * The collected (app, user, day) counts drive the usage-proportional creator
 * payouts described in proappstore.online/pricing.
 *
 * Design notes:
 *
 * - **Visible-time only.** We start a stopwatch when the tab is visible and
 *   pause on `visibilitychange`. A tab that's been hidden for 90s contributes
 *   0 seconds for that interval.
 * - **Browser-only.** All methods no-op when `document` or `window` is
 *   undefined so the import is SSR-safe.
 * - **Silent failures.** Telemetry must never break an app. Every network
 *   path catches + ignores errors.
 * - **Page-close flush.** On `pagehide` we send a final ping with the
 *   residual elapsed seconds via `navigator.sendBeacon` (survives unload)
 *   falling back to `fetch(..., { keepalive: true })`.
 */
const HEARTBEAT_MS = 60_000;
const MAX_DELTA_SECONDS = 90;
function hasBrowser() {
    return typeof document !== 'undefined' && typeof window !== 'undefined';
}
export class Usage {
    appId;
    apiBase;
    auth;
    running = false;
    timer = null;
    visibleSince = null;
    /** Accumulated visible-time since the last successful ping, in milliseconds. */
    accruedMs = 0;
    pendingApiCalls = 0;
    onVisibility = null;
    onPageHide = null;
    constructor(appId, apiBase, auth) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** Begin heartbeat reporting. Idempotent — calling twice is a no-op. */
    start() {
        if (this.running)
            return;
        if (!hasBrowser())
            return;
        this.running = true;
        // Initial visibility snapshot.
        if (document.visibilityState === 'visible') {
            this.visibleSince = performance.now();
        }
        this.onVisibility = () => {
            if (document.visibilityState === 'visible') {
                if (this.visibleSince == null)
                    this.visibleSince = performance.now();
            }
            else {
                this.bankVisible();
            }
        };
        document.addEventListener('visibilitychange', this.onVisibility);
        this.onPageHide = () => {
            this.flush();
        };
        window.addEventListener('pagehide', this.onPageHide);
        this.timer = setInterval(() => {
            this.tick().catch(() => { });
        }, HEARTBEAT_MS);
    }
    /** Stop heartbeats. Idempotent. Doesn't flush; call `flush()` if you need to. */
    stop() {
        if (!this.running)
            return;
        this.running = false;
        if (this.timer) {
            clearInterval(this.timer);
            this.timer = null;
        }
        if (hasBrowser()) {
            if (this.onVisibility)
                document.removeEventListener('visibilitychange', this.onVisibility);
            if (this.onPageHide)
                window.removeEventListener('pagehide', this.onPageHide);
        }
        this.onVisibility = null;
        this.onPageHide = null;
        this.visibleSince = null;
        this.accruedMs = 0;
        this.pendingApiCalls = 0;
    }
    /**
     * Record API calls. Bumps a local counter that piggybacks on the next
     * heartbeat. Cheap to call from hot paths — no network until the next tick.
     */
    recordApiCall(n = 1) {
        if (!Number.isFinite(n) || n <= 0)
            return;
        this.pendingApiCalls += Math.floor(n);
    }
    /**
     * Send a final ping (intended for page-close paths). Best-effort,
     * fire-and-forget. Uses sendBeacon when available so the request survives
     * unload; falls back to keepalive fetch otherwise.
     */
    flush() {
        if (!hasBrowser())
            return;
        this.bankVisible();
        const seconds = this.drainSeconds();
        const apiCalls = this.drainApiCalls();
        if (seconds === 0 && apiCalls === 0)
            return;
        this.send(seconds, apiCalls, /* keepalive */ true);
    }
    // ── internal ──────────────────────────────────────────────────────────────
    /** Add any in-progress visible-time to the running accrual. */
    bankVisible() {
        if (this.visibleSince == null)
            return;
        const now = performance.now();
        this.accruedMs += Math.max(0, now - this.visibleSince);
        this.visibleSince = null;
    }
    /** Round accrued ms to whole seconds, clamp to MAX_DELTA_SECONDS, return + reset. */
    drainSeconds() {
        const whole = Math.floor(this.accruedMs / 1000);
        if (whole <= 0)
            return 0;
        const sent = Math.min(MAX_DELTA_SECONDS, whole);
        this.accruedMs -= sent * 1000;
        // Cap residual so a long-hidden flush after a long visible run doesn't
        // ride along on the next ping forever.
        if (this.accruedMs < 0)
            this.accruedMs = 0;
        return sent;
    }
    drainApiCalls() {
        const n = this.pendingApiCalls;
        this.pendingApiCalls = 0;
        return n;
    }
    async tick() {
        this.bankVisible();
        if (document.visibilityState === 'visible') {
            this.visibleSince = performance.now();
        }
        const seconds = this.drainSeconds();
        const apiCalls = this.drainApiCalls();
        if (seconds === 0 && apiCalls === 0)
            return;
        if (!this.auth.token) {
            // Re-pocket the work — the user may sign in later and we'd like the
            // accrued time to count from then. Note: a session-token signin BEFORE
            // any visible time was banked won't have anything to attribute, and
            // that's fine — anonymous usage isn't part of the payout math anyway.
            this.accruedMs += seconds * 1000;
            this.pendingApiCalls += apiCalls;
            return;
        }
        await this.send(seconds, apiCalls, /* keepalive */ false);
    }
    async send(seconds, apiCalls, keepalive) {
        const token = this.auth.token;
        if (!token)
            return;
        const body = JSON.stringify({
            appId: this.appId,
            deltaSeconds: seconds,
            deltaApiCalls: apiCalls,
        });
        const url = `${this.apiBase}/v1/usage/ping`;
        // Prefer sendBeacon for keepalive (unload survivor) paths. sendBeacon
        // doesn't allow setting an Authorization header, so we encode the token
        // into the URL only for keepalive — server accepts it as a fallback.
        // (If your server doesn't support that, fall through to keepalive fetch.)
        if (keepalive && typeof navigator !== 'undefined' && typeof navigator.sendBeacon === 'function') {
            try {
                const blob = new Blob([body], { type: 'application/json' });
                const ok = navigator.sendBeacon(url, blob);
                if (ok)
                    return;
            }
            catch {
                // fall through
            }
        }
        try {
            await fetch(url, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${token}`,
                },
                body,
                keepalive,
            });
        }
        catch {
            // Telemetry never breaks an app.
        }
    }
}
