// @freeappstore/sdk@19.2.6/dist/auth.js
/**
 * Shared across all FAS apps on the same origin — this is intentional SSO.
 * A user signed in on one FAS app is signed in on all of them.
 */
const STORAGE_KEY = 'fas:session';
/** OAuth authentication — sign in, sign out, session management. */
export class Auth {
    appId;
    apiBase;
    session = null;
    listeners = new Set();
    constructor(appId, apiBase) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.session = this.readStorage();
    }
    /** Current signed-in user, or null if not authenticated. */
    get user() {
        return this.session?.user ?? null;
    }
    /** Current session token, or null if not authenticated. */
    get token() {
        return this.session?.token ?? null;
    }
    /** Subscribe to auth state changes. Fires immediately with current user, then on every change. */
    onChange(listener) {
        this.listeners.add(listener);
        listener(this.user);
        return () => this.listeners.delete(listener);
    }
    /**
     * Redirect-based GitHub OAuth. Opens the platform's hosted OAuth start URL,
     * which redirects back to the current page with a session token in the hash.
     *
     * The current page's `location.hash` is dropped from `return_to` because
     * the OAuth callback writes its own `#fas_session=…` and would clobber any
     * hash-based router state otherwise.
     */
    signIn(provider = 'github') {
        if (typeof window === 'undefined')
            return;
        if (provider === 'email') {
            throw new Error('Use signInWithEmail(email) for email magic-link sign-in.');
        }
        const here = new URL(window.location.href);
        here.hash = '';
        const url = new URL(`/v1/auth/${provider}/start`, this.apiBase);
        url.searchParams.set('app_id', this.appId);
        url.searchParams.set('return_to', here.toString());
        window.location.assign(url.toString());
    }
    /**
     * Email magic-link sign-in. Sends the user an email with a one-time link
     * that completes auth and redirects back here with `#fas_session=…`.
     *
     * Resolves once the email has been queued. The caller should show a
     * "check your inbox" message — the actual sign-in happens later when
     * the user clicks the link.
     *
     * Throws on validation or server errors. Resolves with `{ ok: true }`
     * regardless of whether the email is already registered (no account-
     * enumeration leak).
     */
    async signInWithEmail(email) {
        if (typeof window === 'undefined')
            return;
        const here = new URL(window.location.href);
        here.hash = '';
        const res = await fetch(new URL('/v1/auth/email/start', this.apiBase), {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ email, appId: this.appId, returnTo: here.toString() }),
        });
        if (!res.ok) {
            const body = await res.text();
            throw new Error(`Magic-link request failed: ${res.status} ${body}`);
        }
    }
    /** Clear the session and notify listeners. */
    signOut() {
        this.session = null;
        if (typeof window !== 'undefined') {
            window.localStorage.removeItem(STORAGE_KEY);
        }
        this.emit();
    }
    /**
     * @internal Called by Kv and ApiProxy on 401 responses.
     * Clears the stale session so the UI reacts immediately.
     * Do not call directly — use `signOut()` instead.
     */
    handleUnauthorized() {
        if (this.session)
            this.signOut();
    }
    /**
     * Call this once at app start, before rendering anything that depends on
     * auth state. If the page was loaded via an auth callback (e.g. after
     * `signIn()` returned from GitHub), this captures the session from the
     * URL hash, persists it to localStorage, and clears the hash. On a normal
     * page load it's a no-op — the constructor already restored any cached
     * session from localStorage.
     *
     * @example
     *   const fas = initApp({ appId: 'my-app' });
     *   await fas.auth.init();
     *   render();
     */
    async init() {
        if (typeof window === 'undefined')
            return;
        const hash = window.location.hash;
        if (!hash.startsWith('#fas_session='))
            return;
        // Always clear the hash before doing anything else — even on failure.
        // Otherwise a bad token gets re-tried on every reload and the user is
        // permanently stuck on a "broken" URL.
        history.replaceState(null, '', window.location.pathname + window.location.search);
        let token;
        try {
            token = decodeURIComponent(hash.slice('#fas_session='.length));
        }
        catch {
            // Malformed hash (% with nothing after, etc.). Hash already cleared.
            return;
        }
        if (!token)
            return;
        try {
            const user = await this.fetchUser(token);
            this.session = { token, user };
            this.writeStorage(this.session);
            this.emit();
        }
        catch {
            // Token was invalid or network failed. Hash already cleared so the user
            // won't get stuck in a retry loop. Silently remain signed out.
        }
    }
    /**
     * Set the user's platform-level date of birth. Set-once: throws if it's
     * already set (status 409 from the backend) or if age < 13. After success
     * the cached user is updated and listeners are notified.
     *
     * @param dateOfBirth ISO 'YYYY-MM-DD' string.
     */
    async setDateOfBirth(dateOfBirth) {
        if (!this.session)
            throw new Error('Not signed in.');
        const response = await fetch(new URL('/v1/auth/me/date-of-birth', this.apiBase), {
            method: 'PATCH',
            headers: {
                Authorization: `Bearer ${this.session.token}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({ dateOfBirth }),
        });
        if (response.status === 401) {
            this.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (response.status === 409) {
            throw new Error('Date of birth already set.');
        }
        if (!response.ok) {
            const body = await response.text();
            throw new Error(`setDateOfBirth failed (${response.status}): ${body}`);
        }
        const user = (await response.json());
        this.session = { ...this.session, user };
        this.writeStorage(this.session);
        this.emit();
        return user;
    }
    async fetchUser(token) {
        const response = await fetch(new URL('/v1/auth/me', this.apiBase), {
            headers: { Authorization: `Bearer ${token}` },
        });
        if (!response.ok)
            throw new Error(`Auth failed: ${response.status}`);
        return (await response.json());
    }
    readStorage() {
        if (typeof window === 'undefined')
            return null;
        const raw = window.localStorage.getItem(STORAGE_KEY);
        if (!raw)
            return null;
        try {
            return JSON.parse(raw);
        }
        catch {
            return null;
        }
    }
    writeStorage(session) {
        window.localStorage.setItem(STORAGE_KEY, JSON.stringify(session));
    }
    emit() {
        for (const listener of this.listeners)
            listener(this.user);
    }
}
