// @freeappstore/sdk@19.2.6/dist/db.js
/** Document database — organize data into named collections. */
export class Collections {
    appId;
    apiBase;
    auth;
    constructor(appId, apiBase, auth) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** Get a collection handle for the given name. */
    collection(name) {
        return new Collection(this.appId, name, this.apiBase, this.auth);
    }
}
/** A named collection of JSON documents. */
export class Collection {
    appId;
    name;
    apiBase;
    auth;
    constructor(appId, name, apiBase, auth) {
        this.appId = appId;
        this.name = name;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** Create a new document. Auth required. Returns the document with its generated `id`. */
    async create(doc) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/db/${encodeURIComponent(this.name)}`, this.apiBase);
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(doc),
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok) {
            const text = await response.text();
            throw new Error(`db.create failed (${response.status}): ${text}`);
        }
        return (await response.json());
    }
    /** Get a document by ID. Returns null if not found. Public read. */
    async get(id) {
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/db/${encodeURIComponent(this.name)}/${encodeURIComponent(id)}`, this.apiBase);
        const response = await fetch(url);
        if (response.status === 404)
            return null;
        if (!response.ok)
            throw new Error(`db.get failed: ${response.status}`);
        return (await response.json());
    }
    /** Query documents with optional filtering and pagination. Public read. */
    async query(opts) {
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/db/${encodeURIComponent(this.name)}`, this.apiBase);
        if (opts?.owner)
            url.searchParams.set('owner', opts.owner);
        if (opts?.limit !== undefined)
            url.searchParams.set('limit', String(opts.limit));
        if (opts?.offset !== undefined)
            url.searchParams.set('offset', String(opts.offset));
        if (opts?.orderBy)
            url.searchParams.set('orderBy', opts.orderBy);
        if (opts?.order)
            url.searchParams.set('order', opts.order);
        const response = await fetch(url);
        if (!response.ok)
            throw new Error(`db.query failed: ${response.status}`);
        return (await response.json());
    }
    /** Update a document by merging the patch into the existing data. Auth required, owner only. */
    async update(id, patch) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/db/${encodeURIComponent(this.name)}/${encodeURIComponent(id)}`, this.apiBase);
        const response = await fetch(url, {
            method: 'PUT',
            headers: {
                Authorization: `Bearer ${token}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(patch),
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok) {
            const text = await response.text();
            throw new Error(`db.update failed (${response.status}): ${text}`);
        }
        return (await response.json());
    }
    /** Delete a document. Auth required, owner only. */
    async delete(id) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/db/${encodeURIComponent(this.name)}/${encodeURIComponent(id)}`, this.apiBase);
        const response = await fetch(url, {
            method: 'DELETE',
            headers: { Authorization: `Bearer ${token}` },
        });
        if (response.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!response.ok && response.status !== 404) {
            const text = await response.text();
            throw new Error(`db.delete failed (${response.status}): ${text}`);
        }
    }
}
