// @freeappstore/sdk@19.2.6/dist/index.js
import { Auth } from './auth.js';
import { Counters } from './counters.js';
import { Collections } from './db.js';
import { Kv } from './kv.js';
import { ApiProxy } from './proxy.js';
import { Rooms } from './rooms.js';
export { Collections } from './db.js';
/** Root SDK instance — provides auth, kv, collections, counters, rooms, and proxy sub-clients. */
export class FreeAppStore {
    auth;
    kv;
    collections;
    counters;
    rooms;
    proxy;
    constructor(opts) {
        const apiBase = opts.apiBase ?? 'https://api.freeappstore.online';
        this.auth = new Auth(opts.appId, apiBase);
        this.kv = new Kv(opts.appId, apiBase, this.auth);
        this.collections = new Collections(opts.appId, apiBase, this.auth);
        this.counters = new Counters(opts.appId, apiBase, this.auth);
        this.rooms = new Rooms(opts.appId, apiBase, this.auth);
        this.proxy = new ApiProxy(opts.appId, apiBase, this.auth);
    }
}
/** Create a new FreeAppStore SDK instance for the given app. */
export function initApp(opts) {
    return new FreeAppStore(opts);
}
