// @freeappstore/sdk@19.2.6/dist/kv.js
/**
 * Per-user key-value store, scoped to (appId, userId).
 *
 * Limits (enforced server-side):
 * - max 1MB total per user
 * - max 100 keys per user
 * - max 64KB per value
 *
 * Keys are non-empty strings ≤ 128 chars. We validate client-side so the
 * server doesn't have to deal with edge-case URLs like `/kv/` or absurdly
 * long path segments.
 */
const MAX_KEY_LENGTH = 128;
function assertValidKey(key) {
    if (typeof key !== 'string' || key.length === 0) {
        throw new Error('kv key must be a non-empty string.');
    }
    if (key.length > MAX_KEY_LENGTH) {
        throw new Error(`kv key exceeds ${MAX_KEY_LENGTH} chars.`);
    }
}
export class Kv {
    appId;
    apiBase;
    auth;
    constructor(appId, apiBase, auth) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    /** List all keys for this user. Optionally filter by prefix. */
    async list(opts) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/kv`, this.apiBase);
        if (opts?.prefix)
            url.searchParams.set('prefix', opts.prefix);
        const listResponse = await fetch(url, {
            headers: { Authorization: `Bearer ${token}` },
            ...(opts?.signal && { signal: opts.signal }),
        });
        if (listResponse.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        if (!listResponse.ok)
            throw new Error(`kv.list failed: ${listResponse.status}`);
        let keys = (await listResponse.json());
        // Client-side prefix filter as fallback if server doesn't support it yet
        if (opts?.prefix)
            keys = keys.filter((k) => k.startsWith(opts.prefix));
        return keys;
    }
    /** Fetch multiple keys in parallel. Returns a Map of found key-value pairs (skips failures and missing keys). */
    async getMany(keys, opts) {
        for (const key of keys)
            assertValidKey(key);
        const results = new Map();
        const settled = await Promise.allSettled(keys.map(async (key) => {
            const stored = await this.get(key, opts);
            if (stored !== null)
                results.set(key, stored);
        }));
        // If ALL requests failed with auth error, re-throw so callers know
        const allFailed = settled.every((s) => s.status === 'rejected');
        if (allFailed && settled.length > 0) {
            throw settled[0].reason;
        }
        return results;
    }
    /** Get a value by key. Returns null if not found. */
    async get(key, opts) {
        assertValidKey(key);
        const getResponse = await this.request('GET', key, undefined, opts?.signal);
        if (getResponse.status === 404)
            return null;
        if (!getResponse.ok)
            throw new Error(`kv.get failed: ${getResponse.status}`);
        return (await getResponse.json());
    }
    /** Store a JSON-serializable value under the given key. */
    async set(key, value, opts) {
        assertValidKey(key);
        // JSON.stringify(undefined) returns undefined, which would store an empty
        // body and break later get() calls. Reject up front instead.
        if (value === undefined) {
            throw new Error('kv.set: value is undefined. Use kv.delete(key) to remove a key.');
        }
        const setResponse = await this.request('PUT', key, JSON.stringify(value), opts?.signal);
        if (!setResponse.ok) {
            const text = await setResponse.text();
            throw new Error(`kv.set failed (${setResponse.status}): ${text}`);
        }
    }
    /** Delete a key. No-op if the key doesn't exist. */
    async delete(key, opts) {
        assertValidKey(key);
        const deleteResponse = await this.request('DELETE', key, undefined, opts?.signal);
        if (!deleteResponse.ok && deleteResponse.status !== 404) {
            throw new Error(`kv.delete failed: ${deleteResponse.status}`);
        }
    }
    async request(method, key, body, signal) {
        const token = this.auth.token;
        if (!token)
            throw new Error('Not signed in.');
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/kv/${encodeURIComponent(key)}`, this.apiBase);
        const headers = { Authorization: `Bearer ${token}` };
        if (body !== undefined)
            headers['Content-Type'] = 'application/json';
        const init = { method, headers, ...(signal && { signal }) };
        if (body !== undefined)
            init.body = body;
        const kvResponse = await fetch(url, init);
        if (kvResponse.status === 401) {
            this.auth.handleUnauthorized();
            throw new Error('Not signed in.');
        }
        return kvResponse;
    }
}
