// @freeappstore/sdk@19.2.6/dist/rooms.js
export class Rooms {
    appId;
    apiBase;
    auth;
    constructor(appId, apiBase, auth) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
    }
    join(roomId) {
        return new Room(this.appId, this.apiBase, this.auth, roomId);
    }
}
const RECONNECT_BASE_MS = 1000;
const RECONNECT_MAX_MS = 30_000;
export class Room {
    appId;
    apiBase;
    auth;
    roomId;
    socket = null;
    listeners = new Set();
    peerListeners = new Set();
    stateListeners = new Set();
    peers = [];
    connectionState = 'connecting';
    reconnectAttempt = 0;
    reconnectTimer = null;
    explicitlyClosed = false;
    constructor(appId, apiBase, auth, roomId) {
        this.appId = appId;
        this.apiBase = apiBase;
        this.auth = auth;
        this.roomId = roomId;
        this.connect();
    }
    /** Current connection state. */
    get state() {
        return this.connectionState;
    }
    send(data) {
        if (!this.socket || this.socket.readyState !== WebSocket.OPEN)
            return;
        this.socket.send(JSON.stringify({ kind: 'msg', data }));
    }
    onMessage(listener) {
        this.listeners.add(listener);
        return () => this.listeners.delete(listener);
    }
    onPeers(listener) {
        this.peerListeners.add(listener);
        listener(this.peers);
        return () => this.peerListeners.delete(listener);
    }
    onConnectionState(listener) {
        this.stateListeners.add(listener);
        listener(this.connectionState);
        return () => this.stateListeners.delete(listener);
    }
    /** Permanently close the room. Stops any pending reconnect. */
    close() {
        this.explicitlyClosed = true;
        if (this.reconnectTimer !== null) {
            clearTimeout(this.reconnectTimer);
            this.reconnectTimer = null;
        }
        this.socket?.close();
        this.socket = null;
        this.setState('closed');
        this.listeners.clear();
        this.peerListeners.clear();
        this.stateListeners.clear();
    }
    connect() {
        const token = this.auth.token;
        if (!token) {
            // Auth state may have changed (sign-out). Stop trying.
            this.setState('closed');
            return;
        }
        this.setState('connecting');
        const url = new URL(`/v1/apps/${encodeURIComponent(this.appId)}/rooms/${encodeURIComponent(this.roomId)}`, this.apiBase);
        url.protocol = url.protocol === 'https:' ? 'wss:' : 'ws:';
        url.searchParams.set('token', token);
        const socket = new WebSocket(url.toString());
        this.socket = socket;
        socket.addEventListener('open', () => {
            this.reconnectAttempt = 0;
            this.setState('open');
        });
        socket.addEventListener('message', (ev) => {
            try {
                const parsed = JSON.parse(ev.data);
                if (parsed.kind === 'msg') {
                    for (const l of this.listeners) {
                        l({ from: parsed.from, data: parsed.data, at: parsed.at });
                    }
                }
                else if (parsed.kind === 'peers') {
                    this.peers = parsed.peers;
                    for (const l of this.peerListeners)
                        l(this.peers);
                }
            }
            catch {
                // Ignore malformed frames — server should never send them; if it
                // does, dropping is the right move and an error frame would have
                // come through `kind: 'error'` instead.
            }
        });
        socket.addEventListener('close', () => {
            // Only one of close/error fires the reconnect; we use close because
            // it always fires, even after an error, and is the canonical signal.
            if (this.socket === socket)
                this.socket = null;
            if (this.explicitlyClosed)
                return;
            this.setState('closed');
            this.scheduleReconnect();
        });
        socket.addEventListener('error', () => {
            // We let the close handler do the actual reconnect logic. error is
            // informational and may or may not be followed by close (it always is
            // in browsers per spec).
            this.setState('error');
        });
    }
    scheduleReconnect() {
        if (this.reconnectTimer !== null)
            return;
        // Exponential backoff capped at 30s, with up to 1s of jitter so a
        // backend hiccup doesn't produce a thundering herd of reconnects.
        const backoff = Math.min(RECONNECT_MAX_MS, RECONNECT_BASE_MS * 2 ** this.reconnectAttempt);
        const jitter = Math.random() * 1000;
        this.reconnectAttempt++;
        this.reconnectTimer = setTimeout(() => {
            this.reconnectTimer = null;
            if (this.explicitlyClosed)
                return;
            this.connect();
        }, backoff + jitter);
    }
    setState(state) {
        if (this.connectionState === state)
            return;
        this.connectionState = state;
        for (const l of this.stateListeners)
            l(state);
    }
}
