// @freeappstore/sdk@19.2.6/dist/ui.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, useEffect, useCallback, useRef } from 'react';
import { useAuth, useTheme } from './hooks.js';
/** GitHub avatar with fallback to colored initial circle. */
export function Avatar({ user, size = 32 }) {
    if (user?.avatarUrl) {
        return (_jsx("img", { src: user.avatarUrl, alt: user.login, width: size, height: size, style: {
                borderRadius: '50%',
                display: 'block',
            } }));
    }
    const initial = user?.login?.charAt(0).toUpperCase() ?? '?';
    return (_jsx("div", { style: {
            width: size,
            height: size,
            borderRadius: '50%',
            background: 'var(--accent, #2563eb)',
            color: '#fff',
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            fontSize: size * 0.45,
            fontWeight: 700,
        }, children: initial }));
}
/** Platform-branded sign-in button. Calls app.auth.signIn(). */
export function SignInButton({ app, label = 'Sign in with GitHub' }) {
    return (_jsx("button", { onClick: () => app.auth.signIn(), style: {
            background: 'var(--accent, #2563eb)',
            color: '#fff',
            border: 'none',
            padding: '0.6rem 1.5rem',
            borderRadius: 'var(--radius, 0.75rem)',
            fontSize: '0.9rem',
            fontWeight: 700,
            cursor: 'pointer',
            fontFamily: 'inherit',
        }, children: label }));
}
// ---------------------------------------------------------------------------
// ThemeToggle
// ---------------------------------------------------------------------------
/** Sun/moon toggle button. Cycles: system -> light -> dark. */
export function ThemeToggle() {
    const { theme, preference, setPreference } = useTheme();
    const cycle = useCallback(() => {
        const order = ['system', 'light', 'dark'];
        const idx = order.indexOf(preference);
        setPreference(order[(idx + 1) % order.length]);
    }, [preference, setPreference]);
    // Sun icon for dark mode, moon for light
    const icon = theme === 'dark' ? (
    // Sun
    _jsxs("svg", { width: "18", height: "18", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round", children: [_jsx("circle", { cx: "12", cy: "12", r: "5" }), _jsx("line", { x1: "12", y1: "1", x2: "12", y2: "3" }), _jsx("line", { x1: "12", y1: "21", x2: "12", y2: "23" }), _jsx("line", { x1: "4.22", y1: "4.22", x2: "5.64", y2: "5.64" }), _jsx("line", { x1: "18.36", y1: "18.36", x2: "19.78", y2: "19.78" }), _jsx("line", { x1: "1", y1: "12", x2: "3", y2: "12" }), _jsx("line", { x1: "21", y1: "12", x2: "23", y2: "12" }), _jsx("line", { x1: "4.22", y1: "19.78", x2: "5.64", y2: "18.36" }), _jsx("line", { x1: "18.36", y1: "5.64", x2: "19.78", y2: "4.22" })] })) : (
    // Moon
    _jsx("svg", { width: "18", height: "18", viewBox: "0 0 24 24", fill: "none", stroke: "currentColor", strokeWidth: "2", strokeLinecap: "round", strokeLinejoin: "round", children: _jsx("path", { d: "M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z" }) }));
    return (_jsx("button", { onClick: cycle, "aria-label": `Theme: ${preference}`, title: `Theme: ${preference}`, style: {
            width: 36,
            height: 36,
            borderRadius: 'var(--radius, 0.75rem)',
            border: '1px solid var(--border, #e2e8f0)',
            background: 'var(--surface, #ffffff)',
            color: 'var(--ink, #1e293b)',
            display: 'inline-flex',
            alignItems: 'center',
            justifyContent: 'center',
            cursor: 'pointer',
            padding: 0,
            fontFamily: 'inherit',
        }, children: icon }));
}
/** Avatar button that opens dropdown: username, theme toggle, sign out, delete account. */
export function ProfileMenu({ app, showThemeToggle = true, children }) {
    const { user, signOut, deleteAccount } = useAuth(app);
    const [open, setOpen] = useState(false);
    const ref = useRef(null);
    // Close on outside click
    useEffect(() => {
        if (!open)
            return;
        const handler = (e) => {
            if (ref.current && !ref.current.contains(e.target))
                setOpen(false);
        };
        document.addEventListener('mousedown', handler);
        return () => document.removeEventListener('mousedown', handler);
    }, [open]);
    if (!user)
        return null;
    const handleSignOut = () => { signOut(); setOpen(false); };
    const handleDelete = async () => {
        if (!confirm('Delete your account? This permanently removes ALL your data across ALL apps. This cannot be undone.'))
            return;
        if (!confirm('Are you absolutely sure? Last chance.'))
            return;
        await deleteAccount();
        setOpen(false);
    };
    return (_jsxs("div", { ref: ref, style: { position: 'relative' }, children: [_jsx("button", { onClick: () => setOpen(!open), style: {
                    background: 'none',
                    border: '2px solid var(--border, #e2e8f0)',
                    borderRadius: '50%',
                    padding: 0,
                    cursor: 'pointer',
                    width: 32,
                    height: 32,
                    overflow: 'hidden',
                    display: 'block',
                }, children: _jsx(Avatar, { user: user, size: 28 }) }), open && (_jsxs("div", { style: {
                    position: 'absolute',
                    top: 40,
                    right: 0,
                    background: 'var(--surface, #ffffff)',
                    border: '1px solid var(--border, #e2e8f0)',
                    borderRadius: 'var(--radius, 0.75rem)',
                    boxShadow: '0 4px 20px rgba(0,0,0,0.1)',
                    minWidth: 200,
                    padding: '0.5rem 0',
                    zIndex: 100,
                }, children: [_jsx("div", { style: {
                            padding: '0.5rem 1rem',
                            borderBottom: '1px solid var(--border, #e2e8f0)',
                            fontSize: '0.85rem',
                            fontWeight: 700,
                            color: 'var(--ink, #1e293b)',
                        }, children: user.login }), showThemeToggle && (_jsxs("div", { style: { padding: '0.5rem 1rem', borderBottom: '1px solid var(--border, #e2e8f0)', display: 'flex', alignItems: 'center', justifyContent: 'space-between' }, children: [_jsx("span", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)' }, children: "Theme" }), _jsx(ThemeToggle, {})] })), children, _jsx("button", { onClick: handleSignOut, style: menuItemStyle, children: "Sign out" }), _jsx("button", { onClick: handleDelete, style: { ...menuItemStyle, color: '#dc2626' }, children: "Delete account" })] }))] }));
}
const menuItemStyle = {
    display: 'block',
    width: '100%',
    padding: '0.5rem 1rem',
    background: 'none',
    border: 'none',
    textAlign: 'left',
    fontSize: '0.85rem',
    cursor: 'pointer',
    color: 'var(--ink, #1e293b)',
    fontFamily: 'inherit',
};
/** Full-page settings view: avatar, username, theme selector, sign out, delete account. */
export function ProfilePage({ app, showThemeToggle = true }) {
    const { user, loading, signIn, signOut, deleteAccount } = useAuth(app);
    const { preference, setPreference } = useTheme();
    if (loading) {
        return _jsx("div", { style: { padding: '2rem', textAlign: 'center', color: 'var(--muted, #64748b)' }, children: "Loading..." });
    }
    if (!user) {
        return (_jsxs("div", { style: { padding: '2rem', textAlign: 'center' }, children: [_jsx("p", { style: { color: 'var(--muted, #64748b)', marginBottom: '1rem' }, children: "Sign in to view your profile." }), _jsx(SignInButton, { app: app })] }));
    }
    const handleDelete = async () => {
        if (!confirm('Delete your account? This permanently removes ALL your data across ALL apps. This cannot be undone.'))
            return;
        if (!confirm('Are you absolutely sure? Last chance.'))
            return;
        await deleteAccount();
    };
    const themeOptions = [
        { value: 'system', label: 'System' },
        { value: 'light', label: 'Light' },
        { value: 'dark', label: 'Dark' },
    ];
    return (_jsxs("div", { style: { maxWidth: 480, margin: '0 auto', padding: '2rem 1rem' }, children: [_jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '1rem', marginBottom: '2rem' }, children: [_jsx(Avatar, { user: user, size: 64 }), _jsxs("div", { children: [_jsx("div", { style: { fontSize: '1.25rem', fontWeight: 700, color: 'var(--ink-strong, var(--ink, #0f172a))' }, children: user.login }), _jsx("div", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)' }, children: "FreeAppStore account" })] })] }), showThemeToggle && (_jsxs("div", { style: {
                    background: 'var(--surface, #ffffff)',
                    border: '1px solid var(--border, #e2e8f0)',
                    borderRadius: 'var(--radius, 0.75rem)',
                    padding: '1.25rem',
                    marginBottom: '1rem',
                }, children: [_jsx("div", { style: { fontSize: '0.9rem', fontWeight: 700, marginBottom: '0.75rem', color: 'var(--ink, #1e293b)' }, children: "Appearance" }), _jsx("div", { style: { display: 'flex', gap: '0.5rem' }, children: themeOptions.map((opt) => (_jsx("button", { onClick: () => setPreference(opt.value), style: {
                                flex: 1,
                                padding: '0.5rem',
                                borderRadius: 'var(--radius-sm, 0.5rem)',
                                border: preference === opt.value ? '2px solid var(--accent, #2563eb)' : '1px solid var(--border, #e2e8f0)',
                                background: preference === opt.value ? 'var(--accent-soft, #eff6ff)' : 'transparent',
                                color: preference === opt.value ? 'var(--accent, #2563eb)' : 'var(--muted, #64748b)',
                                fontWeight: preference === opt.value ? 700 : 500,
                                fontSize: '0.85rem',
                                cursor: 'pointer',
                                fontFamily: 'inherit',
                            }, children: opt.label }, opt.value))) })] })), _jsx("button", { onClick: signOut, style: {
                    width: '100%',
                    padding: '0.75rem',
                    borderRadius: 'var(--radius, 0.75rem)',
                    border: '1px solid var(--border, #e2e8f0)',
                    background: 'var(--surface, #ffffff)',
                    color: 'var(--ink, #1e293b)',
                    fontSize: '0.9rem',
                    fontWeight: 600,
                    cursor: 'pointer',
                    marginBottom: '1.5rem',
                    fontFamily: 'inherit',
                }, children: "Sign out" }), _jsxs("div", { style: {
                    border: '1px solid #fecaca',
                    borderRadius: 'var(--radius, 0.75rem)',
                    padding: '1.25rem',
                }, children: [_jsx("div", { style: { fontSize: '0.9rem', fontWeight: 700, color: '#dc2626', marginBottom: '0.5rem' }, children: "Danger zone" }), _jsx("p", { style: { fontSize: '0.85rem', color: 'var(--muted, #64748b)', marginBottom: '0.75rem' }, children: "Permanently delete your account and all data across all apps." }), _jsx("button", { onClick: handleDelete, style: {
                            padding: '0.5rem 1rem',
                            borderRadius: 'var(--radius-sm, 0.5rem)',
                            border: '1px solid #dc2626',
                            background: 'transparent',
                            color: '#dc2626',
                            fontSize: '0.85rem',
                            fontWeight: 600,
                            cursor: 'pointer',
                            fontFamily: 'inherit',
                        }, children: "Delete account" })] })] }));
}
/**
 * Full wrapper: sticky topbar (logo + app name + ProfileMenu or SignInButton),
 * main content area, "Part of FreeAppStore" footer. Optional auth gate.
 */
export function FasShell({ app, children, appName, requireAuth = false, showThemeToggle = true }) {
    const { user, loading } = useAuth(app);
    if (loading) {
        return _jsx("div", { style: { minHeight: '100dvh', display: 'flex', alignItems: 'center', justifyContent: 'center' }, children: _jsx("p", { style: { color: 'var(--muted, #64748b)' }, children: "Loading..." }) });
    }
    if (requireAuth && !user) {
        return (_jsx("div", { style: { minHeight: '100dvh', display: 'flex', alignItems: 'center', justifyContent: 'center', padding: '2rem' }, children: _jsxs("div", { style: { maxWidth: 400, textAlign: 'center' }, children: [_jsx("h1", { style: { fontSize: '1.5rem', fontWeight: 800, marginBottom: '0.5rem', color: 'var(--ink, #1e293b)' }, children: appName || 'FreeAppStore' }), _jsx("p", { style: { color: 'var(--muted, #64748b)', fontSize: '0.9rem', marginBottom: '1rem' }, children: "Sign in to continue." }), _jsx(SignInButton, { app: app }), _jsx("p", { style: { color: 'var(--muted, #64748b)', fontSize: '0.75rem', marginTop: '0.75rem' }, children: "One account for all Free apps." })] }) }));
    }
    return (_jsxs("div", { style: { minHeight: '100dvh', display: 'flex', flexDirection: 'column' }, children: [_jsxs("header", { style: {
                    display: 'flex',
                    alignItems: 'center',
                    justifyContent: 'space-between',
                    padding: '0.5rem 1rem',
                    borderBottom: '1px solid var(--border, #e2e8f0)',
                    background: 'var(--surface, #ffffff)',
                    position: 'sticky',
                    top: 0,
                    zIndex: 50,
                }, children: [_jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.75rem' }, children: [_jsx("a", { href: "https://freeappstore.online", style: { fontWeight: 800, fontSize: '1rem', color: 'var(--accent, #2563eb)', textDecoration: 'none' }, children: "Free" }), appName && _jsx("span", { style: { fontSize: '0.85rem', fontWeight: 600, color: 'var(--muted, #64748b)' }, children: appName })] }), _jsxs("div", { style: { display: 'flex', alignItems: 'center', gap: '0.5rem' }, children: [showThemeToggle && !user && _jsx(ThemeToggle, {}), user ? (_jsx(ProfileMenu, { app: app, showThemeToggle: showThemeToggle })) : (_jsx(SignInButton, { app: app, label: "Sign in" }))] })] }), _jsx("main", { style: { flex: 1, display: 'flex', flexDirection: 'column' }, children: children }), _jsxs("footer", { style: {
                    padding: '1rem',
                    textAlign: 'center',
                    fontSize: '0.75rem',
                    color: 'var(--muted, #64748b)',
                    borderTop: '1px solid var(--border, #e2e8f0)',
                }, children: ["Part of", ' ', _jsx("a", { href: "https://freeappstore.online", style: { color: 'var(--accent, #2563eb)', fontWeight: 600, textDecoration: 'none' }, children: "FreeAppStore" })] })] }));
}
